pkg/virtual/initializingworkspaces/builder/caching_authorizer.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package builder
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	kcpkubernetesclientset "github.com/kcp-dev/client-go/kubernetes"
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	"k8s.io/apimachinery/pkg/util/cache"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/klog/v2"
+
+	corev1alpha1 "github.com/kcp-dev/kcp/pkg/apis/core/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/tenancy/initialization"
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/authorization/delegated"
+	dynamiccontext "github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/context"
+)
+
+func newCachingAuthorizer(client kcpkubernetesclientset.ClusterInterface) *cachingAuthorizer {
```

Comment:
am surprised to see this a a private helper. Isn't this a very generic pattern we want to use in many places?

Reply:
I tend to agree, however in its current form the caching authorizer is not reusable. It is hardcoded to assert the `initialize` verb against `workspacetypes`. I rather suggest to either:

a) make the `Authorize` logic pluggable and make this constructor public or
b) leave this private and call it `newInitializingWorkspacesAuthorizer` and leave the caching logic as an internal implementation detail.

Reply:
Personal preference is to go with b) so we can indeed reuse this in other places 👍